third_party/move/move-compiler/src/cfgir/optimize/inline_blocks.rs

```diff
@@ -131,9 +146,18 @@ fn remap_to_last_target(
     if remapping.is_empty() {
         return;
     }
-    // populate remapping for non changed blocks
+    // populate remapping for non changed blocks, and also
+    // close transitive chains (lab1 -> lab2 -> lab3 becomes lab1 -> lab3).
     for label in blocks.keys() {
-        remapping.entry(*label).or_insert(*label);
+        let mut label2 = *label;
+        while let Some(target) = remapping.get(&label2) {
+            // proceed until label2 has no mapping, or is self-mapping.
+            if label2 == *target {
+                break;
+            }
+            label2 = *target;
+        }
+        remapping.entry(*label).or_insert(label2);
```

Comment:
I don't think this closes transitive chains, as suggested in the comments.

Consider the provided example: lab1 -> lab2 -> lab3, when we reach this line for label == lab1, remapping already has an entry for lab1, so the or_insert wont happen, leaving the mapping as is.

Reply:
Thanks, good eye.   I've rewritten it more carefully, and use a missing entry to represent identity.